src/hotspot/share/services/virtualMemoryTracker.cpp

```diff
@@ -671,3 +671,27 @@ bool VirtualMemoryTracker::walk_virtual_memory(VirtualMemoryWalker* walker) {
    }
   return true;
 }
+
+class FindInRegionWalker : public VirtualMemoryWalker {
+private:
+  const address               _p;
+  const ReservedMemoryRegion* _region;
+public:
+  FindInRegionWalker(void* p) : _p((address)p), _region(nullptr) { }
+
+  bool do_allocation_site(const ReservedMemoryRegion* rgn) {
+    if (_p >= rgn->base() && _p < rgn->base() + rgn->size()) {
+      _region = rgn;
+      return false;
+    }
+    return true;
+  }
+
+  const ReservedMemoryRegion* region() const { return _region; }
+};
+
+const ReservedMemoryRegion* VirtualMemoryTracker::find_region(void* p) {
```

Comment:
It is not thread-safe. But the command is running inside debuger, I don't see there is concurrent case.

Reply:
Not sure about gdb, but I tried running a simple multi threaded program inside lldb on the mac. When calling a function, lldb will resume the program so other threads will execute concurrently.

In this case, `walk_virtual_memory` is protected by a `ThreadCritical`. However, the value returned by `find_region` may become invalid as soon as `walk_virtual_memory` returns because it could be deallocated by another thread.

Reply:
Then malloc case is even worse, the block could be deallocated by another thread. Even `oop->print()` is questionable ...

Reply:
Can the printing be done inside do_allocation_site() instead? That way we are holding the ThreadCritical that will prevent concurrent NMT record/remove operations.

Reply:
Yes, we can. But I think dealing with malloc'd memory is more problematic, if call triggers concurrent execution.

I am going to withdraw this for now and take another look how debuggers behave.

Thanks

Reply:
gdb also resumes all threads upon function call.